File: content/problems/3080-mark-elements-on-array-by-performing-queries/solutions/sorted-indices-and-monotone-pointer/solution.cpp
class Solution {
    public:vector<long long> unmarkedSumArray(vector<int>& nums,vector<vector<int>>& queries){
        int n=nums.size(),pointer=0;
        long long remaining=accumulate(nums.begin(),nums.end(),0LL);
        vector<int> order(n);
        iota(order.begin(),order.end(),0);
        sort(order.begin(),order.end(),[&](int a,int b){
            return nums[a]!=nums[b]?nums[a]<nums[b]:a<b;
        });
        vector<char> marked(n);
        vector<long long> answer;
        for(auto& query:queries){
            int index=query[0];
            if(!marked[index]){
                marked[index]=1;
                remaining-=nums[index];
            }
            int need=query[1];
            while(need>0){
                while(pointer<n&&marked[order[pointer]])++pointer;
                if(pointer==n)break;
                int chosen=order[pointer++];
                marked[chosen]=1;
                remaining-=nums[chosen];
                --need;
            }
            answer.push_back(remaining);
        }
        return answer;
    }
};

File: content/problems/3080-mark-elements-on-array-by-performing-queries/solutions/sorted-indices-and-monotone-pointer/Solution.java
class Solution {
    public long[] unmarkedSumArray(int[] nums,int[][] queries){
        int n=nums.length;
        long remaining=0;
        Integer[] order=new Integer[n];
        for(int i=0;i<n;i++){
            remaining+=nums[i];
            order[i]=i;
        }
        Arrays.sort(order,(a,b)->nums[a]!=nums[b]?Integer.compare(nums[a],nums[b]):Integer.compare(a,b));
        boolean[] marked=new boolean[n];
        long[] answer=new long[queries.length];
        int pointer=0;
        for(int query=0;query<queries.length;query++){
            int index=queries[query][0];
            if(!marked[index]){
                marked[index]=true;
                remaining-=nums[index];
            }
            int need=queries[query][1];
            while(need>0){
                while(pointer<n&&marked[order[pointer]])pointer++;
                if(pointer==n)break;
                int chosen=order[pointer++];
                marked[chosen]=true;
                remaining-=nums[chosen];
                need--;
            }
            answer[query]=remaining;
        }
        return answer;
    }
}
